app/services/irs_attempts_api/tracker.rb

```diff
@@ -1,20 +1,32 @@
 module IrsAttemptsApi
   class Tracker
-    attr_reader :session_id, :enabled_for_session
+    attr_reader :session_id, :enabled_for_session, :request, :user, :sp
 
-    def initialize(session_id:, enabled_for_session:)
-      @session_id = session_id
+    def initialize(session_id:, request:, user:, sp:, enabled_for_session:)
+      @session_id = session_id # IRS session ID
+      @request = request
+      @user = user
+      @sp = sp
       @enabled_for_session = enabled_for_session
     end
 
     def track_event(event_type, metadata = {})
       return unless enabled?
 
+      event_metadata = {
+        user_agent: request&.headers['User-Agent'],
+        unique_session_id: hashed_session_id,
+        user_uuid: AgencyIdentityLinker.for(user: user, service_provider: sp),
+        device_uuid: "FIXME", #user&.device&.cookie_uuid,
+        user_ip_address: request&.remote_ip,
+        irs_application_url: "FIXME", #request&.headers['Referer'],
```

Comment:
@zachmargolis Likewise - you mentioned this being stored, but a bit of poking around didn't turn up an obvious way to get the context from the SP out of session/request data.

Reply:
In a controller we have `service_provider_request` so it would be `service_provider_request&.url` and then we extract the params from there so it might be like `service_provider_session_decorator.request_url_params[:redirect_uri]`?